// @contentful/rich-text-html-renderer/dist/lib/__test__/documents/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.tableWithHeaderDoc = exports.tableDoc = exports.quoteDoc = exports.ulDoc = exports.olDoc = exports.embeddedEntryDoc = exports.marksDoc = exports.headingDoc = exports.paragraphDoc = exports.invalidTypeDoc = exports.invalidMarksDoc = exports.hyperlinkDoc = exports.hrDoc = void 0;
var hr_1 = require("./hr");
Object.defineProperty(exports, "hrDoc", { enumerable: true, get: function () { return __importDefault(hr_1).default; } });
var hyperlink_1 = require("./hyperlink");
Object.defineProperty(exports, "hyperlinkDoc", { enumerable: true, get: function () { return __importDefault(hyperlink_1).default; } });
var invalid_marks_1 = require("./invalid-marks");
Object.defineProperty(exports, "invalidMarksDoc", { enumerable: true, get: function () { return __importDefault(invalid_marks_1).default; } });
var invalid_type_1 = require("./invalid-type");
Object.defineProperty(exports, "invalidTypeDoc", { enumerable: true, get: function () { return __importDefault(invalid_type_1).default; } });
var paragraph_1 = require("./paragraph");
Object.defineProperty(exports, "paragraphDoc", { enumerable: true, get: function () { return __importDefault(paragraph_1).default; } });
var heading_1 = require("./heading");
Object.defineProperty(exports, "headingDoc", { enumerable: true, get: function () { return __importDefault(heading_1).default; } });
var mark_1 = require("./mark");
Object.defineProperty(exports, "marksDoc", { enumerable: true, get: function () { return __importDefault(mark_1).default; } });
var embedded_entry_1 = require("./embedded-entry");
Object.defineProperty(exports, "embeddedEntryDoc", { enumerable: true, get: function () { return __importDefault(embedded_entry_1).default; } });
var ol_1 = require("./ol");
Object.defineProperty(exports, "olDoc", { enumerable: true, get: function () { return __importDefault(ol_1).default; } });
var ul_1 = require("./ul");
Object.defineProperty(exports, "ulDoc", { enumerable: true, get: function () { return __importDefault(ul_1).default; } });
var quote_1 = require("./quote");
Object.defineProperty(exports, "quoteDoc", { enumerable: true, get: function () { return __importDefault(quote_1).default; } });
var table_1 = require("./table");
Object.defineProperty(exports, "tableDoc", { enumerable: true, get: function () { return __importDefault(table_1).default; } });
var table_header_1 = require("./table-header");
Object.defineProperty(exports, "tableWithHeaderDoc", { enumerable: true, get: function () { return __importDefault(table_header_1).default; } });
